File: 0037-sudoku-solver/0037-sudoku-solver.py
class Solution(object):
    def isSafe(self,board, row, col,dig):
        #horizontal
        for j in range(9):
            if board[row][j] == str(dig):
                return False

        #vertical
        for i in range(9):
            if board[i][col] == str(dig):
                return False
        
        #in Grids
        srow = (row//3)*3
        scol = (col//3)*3
        for i in range(srow, srow+3):
            for j in range(scol, scol+3):
                if board[i][j] == str(dig):
                    return False
        
        return True

    def helper(self,board, row, col):
        if row == 9:
            return True
        
        nrow = row
        ncol = col + 1
        if ncol == 9:
            ncol = 0
            nrow = row + 1

        if board[row][col] != '.':
            return self.helper(board,nrow,ncol)

        for dig in map(chr, range(ord('1'), ord('9') + 1)):
            if self.isSafe(board, row, col, dig):
                board[row][col] = dig
                if self.helper(board,nrow,ncol):
                    return True
                board[row][col] = '.'
        
        return False


    def solveSudoku(self, board):
        self.helper(board, 0 ,0)


# class Solution(object):
#     def isSafe(self, board, row, col, dig):
#         # horizontal
#         for j in range(9):
#             if board[row][j] == str(dig):
#                 return False

#         # vertical
#         for i in range(9):
#             if board[i][col] == str(dig):
#                 return False

#         # in Grids
#         srow = (row // 3) * 3
#         scol = (col // 3) * 3  # \U0001f6e0 fixed: was wrongly reassigning srow
#         for i in range(srow, srow + 3):
#             for j in range(scol, scol + 3):
#                 if board[i][j] == str(dig):
#                     return False

#         return True

#     def helper(self, board, row, col):
#         if row == 9:
#             return True

#         nrow = row
#         ncol = col + 1
#         if ncol == 9:  #  fixed: it should be 9, not 0
#             ncol = 0
#             nrow = row + 1

#         if board[row][col] != '.':
#             return self.helper(board, nrow, ncol)  #  fixed: return added

#         for dig in map(chr, range(ord('1'), ord('9') + 1)):
#             if self.isSafe(board, row, col, dig):  #  fixed: added self
#                 board[row][col] = dig
#                 if self.helper(board, nrow, ncol):  #  fixed: added self
#                     return True
#                 board[row][col] = '.'

#         return False

#     def solveSudoku(self, board):
#         self.helper(board, 0, 0)  #  fixed: added self

File: 0037-sudoku-solver/0037-sudoku-solver.cpp
class Solution {
public:
    bool isSafe(vector<vector<char>>& board, int row, int col, char dig) {
        // Check row
        for (int j = 0; j < 9; j++) {
            if (board[row][j] == dig)
                return false;
        }

        // Check column
        for (int i = 0; i < 9; i++) {
            if (board[i][col] == dig)
                return false;
        }

        // Check 3x3 box
        int srow = (row / 3) * 3;
        int scol = (col / 3) * 3;
        for (int i = srow; i < srow + 3; i++) {
            for (int j = scol; j < scol + 3; j++) {
                if (board[i][j] == dig)
                    return false;
            }
        }

        return true;
    }

    bool helper(vector<vector<char>>& board, int row, int col) {
        if (row == 9) return true;

        int nrow = row;
        int ncol = col + 1;
        if (ncol == 9) {
            ncol = 0;
            nrow++;
        }

        if (board[row][col] != '.') {
            return helper(board, nrow, ncol);
        }

        for (char dig = '1'; dig <= '9'; dig++) {
            if (isSafe(board, row, col, dig)) {
                board[row][col] = dig;
                if (helper(board, nrow, ncol)) return true;
                board[row][col] = '.';
            }
        }

        return false;
    }

    void solveSudoku(vector<vector<char>>& board) {
        helper(board, 0, 0);
    }
};
